beacon-chain/core/light-client/store.go

```diff
@@ -24,14 +26,130 @@ type Store struct {
 	lastOptimisticUpdate interfaces.LightClientOptimisticUpdate // tracks the best optimistic update seen so far
 	p2p                  p2p.Accessor
 	stateFeed            event.SubscriberSender
+	cache                *cache // non finality cache
 }
 
-func NewLightClientStore(db iface.HeadAccessDatabase, p p2p.Accessor, e event.SubscriberSender) *Store {
+func NewLightClientStore(ctx context.Context, p p2p.Accessor, e event.SubscriberSender, db iface.HeadAccessDatabase) (*Store, error) {
+	cp, err := db.FinalizedCheckpoint(ctx)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to get finalized checkpoint from database")
+		return nil, nil
+	}
 	return &Store{
 		beaconDB:  db,
 		p2p:       p,
 		stateFeed: e,
+		cache:     newLightClientCache([32]byte(cp.Root)),
+	}, nil
+}
+
+func (s *Store) SaveLCData(ctx context.Context,
+	state state.BeaconState,
+	block interfaces.ReadOnlySignedBeaconBlock,
+	attestedState state.BeaconState,
+	attestedBlock interfaces.ReadOnlySignedBeaconBlock,
+	finalizedBlock interfaces.ReadOnlySignedBeaconBlock,
+	headBlockRoot [32]byte) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// compute required data
+	update, err := NewLightClientUpdateFromBeaconState(ctx, state, block, attestedState, attestedBlock, finalizedBlock)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create light client update")
+	}
+	finalityUpdate, err := NewLightClientFinalityUpdateFromBeaconState(ctx, state, block, attestedState, attestedBlock, finalizedBlock)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create light client finality update")
+	}
+	optimisticUpdate, err := NewLightClientOptimisticUpdateFromBeaconState(ctx, state, block, attestedState, attestedBlock)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create light client optimistic update")
+	}
+	period := slots.SyncCommitteePeriod(slots.ToEpoch(update.AttestedHeader().Beacon().Slot))
+	blockRoot, err := attestedBlock.Block().HashTreeRoot()
+	if err != nil {
+		return errors.Wrapf(err, "failed to compute attested block root")
+	}
+	parentRoot := [32]byte(update.AttestedHeader().Beacon().ParentRoot)
+	signatureBlockRoot, err := block.Block().HashTreeRoot()
+	if err != nil {
+		return errors.Wrapf(err, "failed to compute signature block root")
 	}
+
+	newBlockIsHead := signatureBlockRoot == headBlockRoot
+
+	// create the new cache item
+	newCacheItem := &cacheItem{
```

Comment:
Gut feels tells me we shouldn't be manipulating the cache manually this way, there should be functions on the cache to get and set values